File: novasight/runtime/kalman.py
from __future__ import annotations

from dataclasses import dataclass
from math import exp, isfinite, sqrt


@dataclass(frozen=True)
class KalmanConfig:
    enabled: bool = True
    acceleration_noise: float = 1200.0
    measurement_noise_x: float = 16.0
    measurement_noise_y: float = 16.0
    max_predict_missing_ms: float = 80.0
    max_predict_steps: int = 5
    max_predict_dt_ms: float = 35.0
    max_position_sigma_px: float = 45.0
    max_covariance_trace: float = 5000.0
    nis_threshold: float = 9.21
    nis_hard_reject: float = 16.0
    min_identity_confidence: float = 0.70
    min_prediction_confidence: float = 0.35
    prediction_decay_tau_ms: float = 45.0


@dataclass(frozen=True)
class EstimatedState:
    track_id: int
    state_ts_ns: int
    x: float
    y: float
    vx: float
    vy: float
    cov_trace: float
    position_sigma_px: float
    nis: float
    prediction_confidence: float
    predicted: bool
    prediction_steps: int
    valid: bool
    reason: str = ""


class KalmanEstimator:
    def __init__(
        self,
        *,
        track_id: int,
        x: float,
        y: float,
        ts_ns: int,
        config: KalmanConfig | None = None,
        identity_confidence: float = 1.0,
    ) -> None:
        self.track_id = int(track_id)
        self.config = config or KalmanConfig()
        self.x = [float(x), float(y), 0.0, 0.0]
        pos_noise_x = max(1.0, float(self.config.measurement_noise_x))
        pos_noise_y = max(1.0, float(self.config.measurement_noise_y))
        self.p = [
            [pos_noise_x, 0.0, 0.0, 0.0],
            [0.0, pos_noise_y, 0.0, 0.0],
            [0.0, 0.0, 1000.0, 0.0],
            [0.0, 0.0, 0.0, 1000.0],
        ]
        self.last_state_ts_ns = int(ts_ns)
        self.last_measurement_ts_ns = int(ts_ns)
        self.last_nis = 0.0
        self.prediction_steps = 0
        self.identity_confidence = _clamp01(identity_confidence)
        self.last_estimate = self._estimate(
            ts_ns=int(ts_ns),
            predicted=False,
            valid=True,
            reason="initialized",
        )

    def update_config(self, config: KalmanConfig) -> None:
        self.config = config

    def measurement_nis(self, x: float, y: float, ts_ns: int) -> float:
        if not self.config.enabled:
            return 0.0
        if int(ts_ns) == self.last_state_ts_ns:
            return self.measurement_nis_current(x, y)
        dt = self._dt(ts_ns)
        x_pred, p_pred = self._predict_matrices(dt)
        innovation = [float(x) - x_pred[0], float(y) - x_pred[1]]
        s00 = p_pred[0][0] + max(1e-6, float(self.config.measurement_noise_x))
        s01 = p_pred[0][1]
        s10 = p_pred[1][0]
        s11 = p_pred[1][1] + max(1e-6, float(self.config.measurement_noise_y))
        return _mahalanobis_2d(innovation, s00, s01, s10, s11)

    def measurement_nis_current(self, x: float, y: float) -> float:
        return self.measurement_nis_from_position(
            x=x,
            y=y,
            reference_x=self.x[0],
            reference_y=self.x[1],
        )

    def measurement_nis_from_position(
        self,
        *,
        x: float,
        y: float,
        reference_x: float,
        reference_y: float,
    ) -> float:
        if not self.config.enabled:
            return 0.0
        innovation = [float(x) - float(reference_x), float(y) - float(reference_y)]
        s00 = self.p[0][0] + max(1e-6, float(self.config.measurement_noise_x))
        s01 = self.p[0][1]
        s10 = self.p[1][0]
        s11 = self.p[1][1] + max(1e-6, float(self.config.measurement_noise_y))
        return _mahalanobis_2d(innovation, s00, s01, s10, s11)

    def predict_only(
        self,
        *,
        ts_ns: int,
        identity_confidence: float | None = None,
    ) -> EstimatedState:
        if not self.config.enabled:
            return self._estimate(
                ts_ns=int(ts_ns),
                predicted=True,
                valid=True,
                reason="kalman disabled",
            )
        if identity_confidence is not None:
            self.identity_confidence = _clamp01(identity_confidence)
        dt = self._dt(ts_ns)
        self.x, self.p = self._predict_matrices(dt)
        self.last_state_ts_ns = int(ts_ns)
        self.prediction_steps += 1
        estimate = self._estimate(
            ts_ns=int(ts_ns),
            predicted=True,
            valid=True,
            reason="predicted",
        )
        self.last_estimate = estimate
        return estimate

    def update(
        self,
        *,
        measurement_x: float,
        measurement_y: float,
        ts_ns: int,
        identity_confidence: float,
    ) -> EstimatedState:
        self.identity_confidence = _clamp01(identity_confidence)
        if not self.config.enabled:
            self.x[0] = float(measurement_x)
            self.x[1] = float(measurement_y)
            self.last_state_ts_ns = int(ts_ns)
            self.last_measurement_ts_ns = int(ts_ns)
            self.prediction_steps = 0
            estimate = self._estimate(
                ts_ns=int(ts_ns),
                predicted=False,
                valid=True,
                reason="kalman disabled",
            )
            self.last_estimate = estimate
            return estimate

        dt = self._dt(ts_ns)
        x_pred, p_pred = self._predict_matrices(dt)
        innovation = [float(measurement_x) - x_pred[0], float(measurement_y) - x_pred[1]]
        r00 = max(1e-6, float(self.config.measurement_noise_x))
        r11 = max(1e-6, float(self.config.measurement_noise_y))
        s00 = p_pred[0][0] + r00
        s01 = p_pred[0][1]
        s10 = p_pred[1][0]
        s11 = p_pred[1][1] + r11
        nis = _mahalanobis_2d(innovation, s00, s01, s10, s11)
        if not isfinite(nis) or nis > float(self.config.nis_hard_reject):
            self.x = x_pred
            self.p = p_pred
            self.last_state_ts_ns = int(ts_ns)
            self.prediction_steps += 1
            estimate = self._estimate(
                ts_ns=int(ts_ns),
                predicted=True,
                valid=False,
                reason="NIS_REJECT",
                nis=nis,
            )
            self.last_estimate = estimate
            return estimate

        inv_s = _inverse_2x2(s00, s01, s10, s11)
        if inv_s is None:
            estimate = self._estimate(
                ts_ns=int(ts_ns),
                predicted=True,
                valid=False,
                reason="SINGULAR_INNOVATION",
                nis=nis,
            )
            self.last_estimate = estimate
            return estimate

        k = [
            [
                p_pred[row][0] * inv_s[0][col] + p_pred[row][1] * inv_s[1][col]
                for col in range(2)
            ]
            for row in range(4)
        ]
        self.x = [
            x_pred[row] + k[row][0] * innovation[0] + k[row][1] * innovation[1]
            for row in range(4)
        ]
        self.p = [
            [
                p_pred[row][col]
                - k[row][0] * p_pred[0][col]
                - k[row][1] * p_pred[1][col]
                for col in range(4)
            ]
            for row in range(4)
        ]
        self.last_nis = float(nis)
        self.last_state_ts_ns = int(ts_ns)
        self.last_measurement_ts_ns = int(ts_ns)
        self.prediction_steps = 0
        estimate = self._estimate(
            ts_ns=int(ts_ns),
            predicted=False,
            valid=True,
            reason="updated",
            nis=nis,
        )
        self.last_estimate = estimate
        return estimate

    def _dt(self, ts_ns: int) -> float:
        raw = (int(ts_ns) - int(self.last_state_ts_ns)) / 1e9
        if not isfinite(raw) or raw < 0:
            return 0.0
        max_dt = max(0.001, float(self.config.max_predict_dt_ms) / 1000.0)
        return min(raw, max_dt)

    def _predict_matrices(self, dt: float) -> tuple[list[float], list[list[float]]]:
        q = _process_noise(dt, max(1e-6, float(self.config.acceleration_noise)))
        x_pred = [
            self.x[0] + dt * self.x[2],
            self.x[1] + dt * self.x[3],
            self.x[2],
            self.x[3],
        ]
        p = self.p
        fp = [
            [p[0][col] + dt * p[2][col] for col in range(4)],
            [p[1][col] + dt * p[3][col] for col in range(4)],
            list(p[2]),
            list(p[3]),
        ]
        p_pred = [
            [fp[row][0] + dt * fp[row][2], fp[row][1] + dt * fp[row][3], fp[row][2], fp[row][3]]
            for row in range(4)
        ]
        for row in range(4):
            for col in range(4):
                p_pred[row][col] += q[row][col]
        return x_pred, p_pred

    def _estimate(
        self,
        *,
        ts_ns: int,
        predicted: bool,
        valid: bool,
        reason: str,
        nis: float | None = None,
    ) -> EstimatedState:
        cov_trace = sum(float(self.p[i][i]) for i in range(4))
        sigma = sqrt(max(0.0, float(self.p[0][0]) + float(self.p[1][1])) / 2.0)
        current_nis = self.last_nis if nis is None else float(nis)
        missing_ms = max(0.0, (int(ts_ns) - int(self.last_measurement_ts_ns)) / 1e6)
        confidence = self._prediction_confidence(
            missing_ms=missing_ms,
            sigma=sigma,
            nis=current_nis,
        )
        computed_valid = (
            bool(valid)
            and all(isfinite(float(value)) for value in self.x)
            and isfinite(cov_trace)
            and isfinite(sigma)
            and missing_ms <= float(self.config.max_predict_missing_ms)
            and self.prediction_steps <= int(self.config.max_predict_steps)
            and sigma <= float(self.config.max_position_sigma_px)
            and cov_trace <= float(self.config.max_covariance_trace)
            and self.identity_confidence >= float(self.config.min_identity_confidence)
            and current_nis <= float(self.config.nis_threshold)
            and confidence >= float(self.config.min_prediction_confidence)
        )
        return EstimatedState(
            track_id=self.track_id,
            state_ts_ns=int(ts_ns),
            x=float(self.x[0]),
            y=float(self.x[1]),
            vx=float(self.x[2]),
            vy=float(self.x[3]),
            cov_trace=float(cov_trace),
            position_sigma_px=float(sigma),
            nis=float(current_nis),
            prediction_confidence=float(confidence),
            predicted=bool(predicted),
            prediction_steps=int(self.prediction_steps),
            valid=bool(computed_valid),
            reason=reason,
        )

    def _prediction_confidence(self, *, missing_ms: float, sigma: float, nis: float) -> float:
        max_sigma = max(1e-6, float(self.config.max_position_sigma_px))
        nis_threshold = max(1e-6, float(self.config.nis_threshold))
        tau_ms = max(1e-6, float(self.config.prediction_decay_tau_ms))
        covariance_confidence = _clamp01(1.0 - sigma / max_sigma)
        residual_confidence = _clamp01(1.0 - max(0.0, nis) / nis_threshold)
        missing_decay = exp(-max(0.0, missing_ms) / tau_ms)
        return _clamp01(
            self.identity_confidence
            * covariance_confidence
            * residual_confidence
            * missing_decay
        )


def _process_noise(dt: float, acceleration_noise: float) -> list[list[float]]:
    dt2 = dt * dt
    dt3 = dt2 * dt
    dt4 = dt2 * dt2
    scale = acceleration_noise
    return [
        [scale * dt4 / 4.0, 0.0, scale * dt3 / 2.0, 0.0],
        [0.0, scale * dt4 / 4.0, 0.0, scale * dt3 / 2.0],
        [scale * dt3 / 2.0, 0.0, scale * dt2, 0.0],
        [0.0, scale * dt3 / 2.0, 0.0, scale * dt2],
    ]


def _mahalanobis_2d(
    innovation: list[float],
    s00: float,
    s01: float,
    s10: float,
    s11: float,
) -> float:
    det = s00 * s11 - s01 * s10
    if not isfinite(det) or abs(det) < 1e-12:
        return float("inf")
    dx = float(innovation[0])
    dy = float(innovation[1])
    return float(
        (dx * (s11 * dx - s01 * dy) + dy * (-s10 * dx + s00 * dy)) / det
    )


def _inverse_2x2(
    a: float,
    b: float,
    c: float,
    d: float,
) -> list[list[float]] | None:
    det = a * d - b * c
    if not isfinite(det) or abs(det) < 1e-12:
        return None
    inv_det = 1.0 / det
    return [[d * inv_det, -b * inv_det], [-c * inv_det, a * inv_det]]


def _clamp01(value: float) -> float:
    return max(0.0, min(1.0, float(value)))

File: crates/novasight-core/src/tracking/kalman.rs
//! Allocation-free constant-velocity Kalman state used by target association.
//!
//! The filtered position is deliberately limited to identity association.
//! Target scoring and mouse control consume the latest observed aim point so
//! filtering cannot add visible selection or control lag.

use serde::{Deserialize, Serialize};

// Runtime-tunable values and internal safety limits share one canonical
// default here. Persistence reads these defaults through `KalmanConfig`
// instead of copying numeric literals into another crate.
const DEFAULT_ACCELERATION_NOISE: f64 = 1_200.0;
const DEFAULT_MEASUREMENT_NOISE_X: f64 = 16.0;
const DEFAULT_MEASUREMENT_NOISE_Y: f64 = 16.0;
const DEFAULT_MAX_PREDICT_DT_MS: f64 = 35.0;
const DEFAULT_MAX_PREDICT_MISSING_MS: f64 = 80.0;
const DEFAULT_MAX_PREDICT_STEPS: u32 = 5;
const DEFAULT_NIS_THRESHOLD: f64 = 9.21;
const DEFAULT_NIS_HARD_REJECT: f64 = 16.0;
const DEFAULT_MAX_POSITION_SIGMA_PX: f64 = 45.0;
const DEFAULT_MAX_COVARIANCE_TRACE: f64 = 5_000.0;
const DEFAULT_MIN_IDENTITY_CONFIDENCE: f64 = 0.70;
const DEFAULT_MIN_PREDICTION_CONFIDENCE: f64 = 0.35;
const DEFAULT_PREDICTION_DECAY_TAU_MS: f64 = 45.0;

#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
pub struct KalmanConfig {
    pub acceleration_noise: f64,
    pub measurement_noise_x: f64,
    pub measurement_noise_y: f64,
    pub max_predict_dt_ms: f64,
    pub max_predict_missing_ms: f64,
    pub max_predict_steps: u32,
    pub nis_threshold: f64,
    pub nis_hard_reject: f64,
    pub max_position_sigma_px: f64,
    pub max_covariance_trace: f64,
    pub min_identity_confidence: f64,
    pub min_prediction_confidence: f64,
    pub prediction_decay_tau_ms: f64,
}

impl Default for KalmanConfig {
    fn default() -> Self {
        Self {
            acceleration_noise: DEFAULT_ACCELERATION_NOISE,
            measurement_noise_x: DEFAULT_MEASUREMENT_NOISE_X,
            measurement_noise_y: DEFAULT_MEASUREMENT_NOISE_Y,
            max_predict_dt_ms: DEFAULT_MAX_PREDICT_DT_MS,
            max_predict_missing_ms: DEFAULT_MAX_PREDICT_MISSING_MS,
            max_predict_steps: DEFAULT_MAX_PREDICT_STEPS,
            nis_threshold: DEFAULT_NIS_THRESHOLD,
            nis_hard_reject: DEFAULT_NIS_HARD_REJECT,
            max_position_sigma_px: DEFAULT_MAX_POSITION_SIGMA_PX,
            max_covariance_trace: DEFAULT_MAX_COVARIANCE_TRACE,
            min_identity_confidence: DEFAULT_MIN_IDENTITY_CONFIDENCE,
            min_prediction_confidence: DEFAULT_MIN_PREDICTION_CONFIDENCE,
            prediction_decay_tau_ms: DEFAULT_PREDICTION_DECAY_TAU_MS,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub(super) struct KalmanState {
    state: [f64; 4],
    covariance: [[f64; 4]; 4],
    state_ts_ns: u64,
    last_measurement_ts_ns: u64,
    last_nis: f64,
    prediction_steps: u32,
    estimate_valid: bool,
    initialized: bool,
}

impl Default for KalmanState {
    fn default() -> Self {
        Self {
            state: [0.0; 4],
            covariance: [[0.0; 4]; 4],
            state_ts_ns: 0,
            last_measurement_ts_ns: 0,
            last_nis: 0.0,
            prediction_steps: 0,
            estimate_valid: false,
            initialized: false,
        }
    }
}

impl KalmanState {
    pub(super) fn new(
        x: f64,
        y: f64,
        ts_ns: u64,
        config: KalmanConfig,
        identity_confidence: f64,
    ) -> Self {
        let mut covariance = [[0.0; 4]; 4];
        covariance[0][0] = config.measurement_noise_x.max(1.0);
        covariance[1][1] = config.measurement_noise_y.max(1.0);
        covariance[2][2] = 1_000.0;
        covariance[3][3] = 1_000.0;
        let mut state = Self {
            state: [x, y, 0.0, 0.0],
            covariance,
            state_ts_ns: ts_ns,
            last_measurement_ts_ns: ts_ns,
            last_nis: 0.0,
            prediction_steps: 0,
            estimate_valid: false,
            initialized: true,
        };
        state.estimate_valid = state.computed_valid(config, identity_confidence);
        state
    }

    pub(super) fn position(&self) -> (f64, f64) {
        (self.state[0], self.state[1])
    }

    pub(super) fn prediction_valid(&self) -> bool {
        self.estimate_valid
    }

    pub(super) fn predict(
        &mut self,
        ts_ns: u64,
        config: KalmanConfig,
        identity_confidence: f64,
    ) -> bool {
        if !self.initialized || ts_ns <= self.state_ts_ns {
            return self.estimate_valid;
        }
        let raw_dt = ts_ns.saturating_sub(self.state_ts_ns) as f64 / 1_000_000_000.0;
        let dt = raw_dt.min(config.max_predict_dt_ms.max(1.0) / 1_000.0);
        let (state, covariance) = self.predicted(dt, config);
        self.state = state;
        self.covariance = covariance;
        self.state_ts_ns = ts_ns;
        self.prediction_steps = self.prediction_steps.saturating_add(1);
        self.estimate_valid = self.computed_valid(config, identity_confidence);
        self.estimate_valid
    }

    pub(super) fn measurement_nis(&self, x: f64, y: f64, config: KalmanConfig) -> f64 {
        let s00 = self.covariance[0][0] + config.measurement_noise_x.max(1e-6);
        let s01 = self.covariance[0][1];
        let s10 = self.covariance[1][0];
        let s11 = self.covariance[1][1] + config.measurement_noise_y.max(1e-6);
        mahalanobis_2d(x - self.state[0], y - self.state[1], s00, s01, s10, s11)
    }

    pub(super) fn measurement_nis_from_position(
        &self,
        x: f64,
        y: f64,
        reference_x: f64,
        reference_y: f64,
        config: KalmanConfig,
    ) -> f64 {
        let s00 = self.covariance[0][0] + config.measurement_noise_x.max(1e-6);
        let s01 = self.covariance[0][1];
        let s10 = self.covariance[1][0];
        let s11 = self.covariance[1][1] + config.measurement_noise_y.max(1e-6);
        mahalanobis_2d(x - reference_x, y - reference_y, s00, s01, s10, s11)
    }

    pub(super) fn update(
        &mut self,
        x: f64,
        y: f64,
        ts_ns: u64,
        config: KalmanConfig,
        identity_confidence: f64,
    ) -> bool {
        let r00 = config.measurement_noise_x.max(1e-6);
        let r11 = config.measurement_noise_y.max(1e-6);
        let s00 = self.covariance[0][0] + r00;
        let s01 = self.covariance[0][1];
        let s10 = self.covariance[1][0];
        let s11 = self.covariance[1][1] + r11;
        let nis = mahalanobis_2d(x - self.state[0], y - self.state[1], s00, s01, s10, s11);
        if !nis.is_finite() || nis > config.nis_hard_reject {
            *self = Self::new(x, y, ts_ns, config, identity_confidence);
            return self.estimate_valid;
        }
        let Some(inverse) = inverse_2x2(s00, s01, s10, s11) else {
            *self = Self::new(x, y, ts_ns, config, identity_confidence);
            return self.estimate_valid;
        };
        let innovation = [x - self.state[0], y - self.state[1]];
        let mut gain = [[0.0; 2]; 4];
        for (row, gain_row) in gain.iter_mut().enumerate() {
            gain_row[0] =
                self.covariance[row][0] * inverse[0][0] + self.covariance[row][1] * inverse[1][0];
            gain_row[1] =
                self.covariance[row][0] * inverse[0][1] + self.covariance[row][1] * inverse[1][1];
        }
        for (row, value) in self.state.iter_mut().enumerate() {
            *value += gain[row][0] * innovation[0] + gain[row][1] * innovation[1];
        }
        let previous = self.covariance;
        self.covariance = std::array::from_fn(|row| {
            std::array::from_fn(|column| {
                previous[row][column]
                    - gain[row][0] * previous[0][column]
                    - gain[row][1] * previous[1][column]
            })
        });
        self.state_ts_ns = ts_ns;
        self.last_measurement_ts_ns = ts_ns;
        self.last_nis = nis;
        self.prediction_steps = 0;
        self.estimate_valid = self.computed_valid(config, identity_confidence);
        self.estimate_valid
    }

    fn predicted(&self, dt: f64, config: KalmanConfig) -> ([f64; 4], [[f64; 4]; 4]) {
        let state = [
            self.state[0] + dt * self.state[2],
            self.state[1] + dt * self.state[3],
            self.state[2],
            self.state[3],
        ];
        let fp: [[f64; 4]; 4] = std::array::from_fn(|row| {
            std::array::from_fn(|column| match row {
                0 => self.covariance[0][column] + dt * self.covariance[2][column],
                1 => self.covariance[1][column] + dt * self.covariance[3][column],
                _ => self.covariance[row][column],
            })
        });
        let mut covariance = std::array::from_fn(|row| {
            [
                fp[row][0] + dt * fp[row][2],
                fp[row][1] + dt * fp[row][3],
                fp[row][2],
                fp[row][3],
            ]
        });
        let dt2 = dt * dt;
        let dt3 = dt2 * dt;
        let dt4 = dt2 * dt2;
        let q = config.acceleration_noise.max(1e-6);
        covariance[0][0] += q * dt4 / 4.0;
        covariance[0][2] += q * dt3 / 2.0;
        covariance[1][1] += q * dt4 / 4.0;
        covariance[1][3] += q * dt3 / 2.0;
        covariance[2][0] += q * dt3 / 2.0;
        covariance[2][2] += q * dt2;
        covariance[3][1] += q * dt3 / 2.0;
        covariance[3][3] += q * dt2;
        (state, covariance)
    }

    fn computed_valid(&self, config: KalmanConfig, identity_confidence: f64) -> bool {
        let covariance_trace = (0..4)
            .map(|index| self.covariance[index][index])
            .sum::<f64>();
        let sigma = ((self.covariance[0][0] + self.covariance[1][1]).max(0.0) * 0.5).sqrt();
        let missing_ms = self.state_ts_ns.saturating_sub(self.last_measurement_ts_ns) as f64 / 1e6;
        let covariance_confidence =
            (1.0 - sigma / config.max_position_sigma_px.max(1e-6)).clamp(0.0, 1.0);
        let residual_confidence =
            (1.0 - self.last_nis.max(0.0) / config.nis_threshold.max(1e-6)).clamp(0.0, 1.0);
        let missing_decay = (-missing_ms / config.prediction_decay_tau_ms.max(1e-6)).exp();
        let prediction_confidence = identity_confidence.clamp(0.0, 1.0)
            * covariance_confidence
            * residual_confidence
            * missing_decay;
        self.state.iter().all(|value| value.is_finite())
            && covariance_trace.is_finite()
            && covariance_trace <= config.max_covariance_trace
            && sigma.is_finite()
            && sigma <= config.max_position_sigma_px
            && missing_ms <= config.max_predict_missing_ms
            && self.prediction_steps <= config.max_predict_steps
            && identity_confidence >= config.min_identity_confidence
            && self.last_nis <= config.nis_threshold
            && prediction_confidence >= config.min_prediction_confidence
    }
}

fn inverse_2x2(a: f64, b: f64, c: f64, d: f64) -> Option<[[f64; 2]; 2]> {
    let determinant = a * d - b * c;
    (determinant.is_finite() && determinant.abs() >= 1e-12).then_some([
        [d / determinant, -b / determinant],
        [-c / determinant, a / determinant],
    ])
}

fn mahalanobis_2d(dx: f64, dy: f64, a: f64, b: f64, c: f64, d: f64) -> f64 {
    let Some(inverse) = inverse_2x2(a, b, c, d) else {
        return f64::INFINITY;
    };
    dx * (inverse[0][0] * dx + inverse[0][1] * dy) + dy * (inverse[1][0] * dx + inverse[1][1] * dy)
}
